Add networking guide docs (#138119)
elastic/elasticsearch: docs/internal/DistributedArchitectureGuide.md

# Networking

Every elasticsearch node maintains various networking clients and servers,
protocols, and synchronous/asynchronous handling.

## HTTP Transport

Every node has an HTTP Transport Server (simply HTTP Transport) and it is a
single entry point for all external clients (except cross-cluster). Management,
ingestion, search, and everything else passes through HTTP server.

Elasticsearch works over HTTP 1.1 and supports HTTP features such as: TLS,
chunked transfer encoding, content compression, pipelining. Although we try to
be HTTP spec compliant, Elastic is not a webserver. We support GET requests with

> payload (some old proxies might drop content), requests cannot be cached by

Reviewer: Maybe mention that in these cases we also support the same API with a different verb, normally POST, for clients who cannot send a GET-with-body.

middle boxes.

By default HTTP server binds to the first available port in range
9200-9300. There is no connections limit, but there is a limit on payload
size. Default maximum payload is 100mb after compression. It's a very large
number and almost never a good target that client should approach.

Security is not enabled by default, meaning no TLS and authentication. These

> features are available in the x-pack/security module.

Reviewer 2: I would not say "Security is not enabled by default" but instead just say "security features are achieved with separate x-pack modules". We had dedicated effort to ensure [security is on by default](https://github.com/elastic/elasticsearch/issues/80170) and basic security including authc/authz/TLS is in the free tier. It is rather disheartening to say the opposite.

HTTP transport provides two options for content processing: aggregate fully and
incremental. Aggregated content is a preferable choice for a small messages that

> cannot be parsed incrementally (like JSON). But aggregation has drawbacks, it

Reviewer: > cannot be parsed incrementally (like JSON)

Mmm sorta depends what you mean by "incrementally" but I'd say that the SAX-style parsing we do is in fact working incrementally. It doesn't really make sense to start parsing before we've received the whole body, but that's different from saying "cannot".

only few to proceed, until pressure is resolved. This feature protects from
unbounded memory usage and OOMs.

ES supports multiple Content-Types for payload: CBOR, JSON, SMILE, YAML, and
their versioned types. Internally we call it XContentType. When a class
implements `toXContent` that means it can be sent(serialized) over HTTP/REST.

HTTP routing is based on a combination of Method and URI. For example,
`RestCreateIndexAction` handler uses `("PUT", "/{index}")`, where curly braces
indicate path variables. `RestBulkAction` specifies a list of routes

```java
    @Override
    public List<Route> routes() {
        return List.of(
            new Route(POST, "/_bulk"),
            new Route(PUT, "/_bulk"),
            new Route(POST, "/{index}/_bulk"),
            new Route(PUT, "/{index}/_bulk")
        );
    }
```

Every REST handler must be declared in `ActionModule` class in the
`initRestHandlers` method. Plugins that implements `ActionPlugin` can extend
list of handlers through `getRestHandlers` override. Every REST handler should
extend `BaseRestHandler`.

The job of the REST handler is to parse and validate HTTP request and construct

> typed version of request, often Transport request (see Transport section below).

Reviewer: Might be worth pointing out here that (if security is enabled then) authentication happens before the REST handler but authorization happens after (when entering the transport layer).

  -> `Rest{Some}Action`
```

Where `Netty4HttpServerTransport` is a single implementation of
`AbstractHttpServerTransport` that lives in a `transport-netty4` module. And
security module injects SSL and headers validator.

## Transport

`Transport` is an umbrella term for a node-to-node communication. It's a
TCP-based custom binary protocol. Every node in a cluster is a client and server
at the same time. Node-to-node communication never uses HTTP transport.

`Netty4Transport` is the only implementation of TCP transport and initialize
Transport client and server. X-pack/security plugin provides secure version of
transport with TLS and authentication - `SecurityNetty4Transport`.

Once node discovers cluster it will open a pool of connections to every other
node in a cluster, and every other node will open a pool of connections
back. That means a connection between nodes A and B is A->B pool + B->A pool. A
node sends requests only on connections it opens (as a client).

A default connection pool is around 13 connections, a pool has sub-pools of
connections for different purposes: ping, node-state, bulks, etc. Pool structure
is defined in `ConnectionProfile` class.

> ES has resilience to disconnects, frequent reconnects, but in general we assume

Reviewer: > resilience

I don't think we are particularly resilient to frequent reconnects. At least, not without being more specific. ES never behaves _incorrectly_ (e.g. loses data) in the face of network outages but it may become unavailable unless the network is stable.

network should be stable between nodes. Nevertheless, connectivity problems
never stop coming from SDH's and Serverless. Not every cluster has enough
redundancy to survive prolonged network interruptions.

> Request timeouts are discouraged. Transport request can take all the time they

Reviewer: Might be worth saying explicitly that every transport request is guaranteed to receive a response eventually, even without a timeout, and even in the face of network disruption (as long as the response handler has no bugs that can lead to it dropping the listener before it's complete).

fail outstanding requests. Then reconnect the pool again.

There are no retries on the Transport layer. Application layer should decide how
and when to retry. Some requests don't retry, and some retry with
`RetryableAction` or `TransportMasterNodeAction`. We might have retries in the
Transport framework in the future with [#95100](https://github.com/elastic/elasticsearch/issues/95100).

Transport does not support multiplexing. Each transport message has to be
dispatched fully before the next message can be sent. It's recommended to
properly size/chunk messages at application layer to provide fairness of
delivery across many senders. There is no strict guideline for transport message
size. Use your judgement based on: how critical message is, how often it needs
to be delivered, is it necessary to deliver as a single large piece. Almost
always it can be chunked.

`TransportMessage` has a tall family tree. There are simple node-to-node
messages, broadcast messages, messages that has to be acknowledge by master
node, etc. These are important for the transport infrastructure to dispatch
message to the right place and return response back when it's truly successful.

## Snapshots

> Another area of different networking clients is snapshotting. ES supports

Reviewer 2: There are other features such as SAML/JWT metadata reloading, watcher http action, reindex and probably ML related features such as inference that also use HTTP clients. I wonder whether we should mention them here. Or do you see snapshots being the most relevant one for the team?

serves HTTP and Transport servers and clients. The pool size is CPU bound, for
example 4 cores would have 4 threads. These 4 threads would serve hundreds or
thousands of connections.

Event-loop threads serve many connections each, so it's absolutely critical to
not block thread for a long time. Any blocking operation or heavy compute has to
be forked to another thread pool. But forking comes with overhead, doing forking
on every tiny request is a wasted CPU work. As a rule of thumb: don't fork
simple requests that can be served from memory and do not require heavy
computations (seconds), otherwise fork.

Transport threads are monitored by `ThreadWatchdog`. When thread runs single
task longer than 5 seconds warning logs appears. There are many reasons for
slowness, not only lack of forking. For example GC pauses, CPU starvation from
other thread-pools.

### ByteBuf - byte buffers and reference counting

One of the performance edges of netty is controlled memory allocation. Netty
manages byte buffer pools and reuse them heavily. This performance gain comes
with a cost. And cost is reference counting on developer's shoulders. Netty

> reads socket bytes into pooled byte-buffers and passes them to application. Then

Reviewer: Might be worth saying a bit more about how this looks in a heap dump - there's a pool of 1MiB `byte[]` objects which Netty slices into 16kiB pages, not all of which may be in use, so you have to account for this when investigating memory usage.

# Networking

Every elasticsearch node maintains various networking clients and servers,
protocols, and synchronous/asynchronous handling.

## HTTP Transport
The HTTP Transport Server (simply HTTP Transport) is a single entry point for
all external clients (excluding cross-cluster communication). Management,
ingestion, search, and all other external operations pass through the HTTP
server.

Elasticsearch works over HTTP 1.1 and supports features such as TLS, chunked
transfer encoding, content compression, and pipelining. While attempting to
be HTTP spec compliant, Elasticsearch is not a webserver. ES Supports `GET`
requests with a payload (though some old proxies may drop content) and
`POST` for clients unable to send `GET-with-body`. Requests cannot be cached
by middle boxes.

There is no connection limit, but a limit on payload size exists. The default
maximum payload is 100MB after compression. It's a very large number and almost
never a good target that the client should approach.

> Security features, including basic security (authc/authz/TLS) in the free tier,

Reviewer 3: ```suggestion
Security features, including basic security (authentication/authorization/Transport Layer Security) in the free tier,
```
For someone that is not used to working with networking and security (me) it's not obvious what authc/authz/TLS means without googling. If target audience is newcomers and open source community I think it's worth spelling those out.

are achieved with separate x-pack modules.

HTTP transport provides two options for content processing: aggregate fully and
incremental. Aggregated content is a preferable choice for small messages that
do not fit for incremental parsing (e.g., JSON). Aggregation has drawbacks, it
requires more memory, which is reserved until all bytes are received. Concurrent
incomplete requests can lead to unbounded memory growth and potential OOMs.
Large delimited content, such as bulk indexing, which is processed in byte
chunks. It's more complicated for application code, but provides better control
over memory usage.

> Incremental bulk indexing includes a back-pressure feature.When memory pressure

Reviewer: ```suggestion
Incremental bulk indexing includes a back-pressure feature. When memory pressure
```

grows high, reading bytes from TCP sockets is paused for some connections,
allowing only a few to proceed until the pressure is resolved. This mechanism
protects against unbounded memory usage and OutOfMemory errors (OOMs).

ES supports multiple `Content-Types` for the payload, collectively referred to

> internally as `XContentType`: CBOR, JSON, SMILE, YAML, and their versioned types.

Reviewer: We can also handle other content-types, it's just that these are the best-supported ones.

The HTTP server provides two options for content processing: full aggregation
and incremental processing. Aggregated content is a preferable choice for small
messages that do not fit for incremental parsing (e.g., JSON). Aggregation has
drawbacks: it requires more memory, which is reserved until all bytes are
received. Concurrent incomplete requests can lead to unbounded memory growth and
potential OOMs. Large delimited content, such as bulk indexing, which is
processed in byte chunks, provides better control over memory usage but is more
complicated for application code.

Incremental bulk indexing includes a back-pressure feature. See `org.
elasticsearch.index.IndexingPressure`. When memory pressure grows high
(`LOW_WATERMARK`), reading bytes from TCP sockets is paused for some
connections, allowing only a few to proceed until the pressure is resolved.
When memory grows too high (`HIGH_WATERMARK`) bulk items are rejected with 429.
This mechanism protects against unbounded memory usage and `OutOfMemory`
errors (OOMs).

ES supports multiple `Content-Types` for the payload. These are
implementations of `MediaType` interface. A common ones internally called

> `XContentType`: CBOR, JSON, SMILE, YAML, and their versioned types. X-pack

Reviewer 2: Nit
```suggestion
implementations of `MediaType` interface. A common implementation is called
`XContentType`, including CBOR, JSON, SMILE, YAML, and their versioned types. X-pack
```

-> `RestController`
-> `BaseRestHandler`
-> `Rest{Some}Action`
```

`Netty4HttpServerTransport` is a single implementation of
`AbstractHttpServerTransport` from the `transport-netty4`
module. Security module injects TLS and headers validator.

## Transport

Transport is the term for node-to-node communication, utilizing a TCP-based
custom binary protocol. Every node acts as both a client and a server.
Node-to-node communication almost never uses HTTP transport (except for
reindex-from-remote).

`Netty4Transport` is the sole implementation of TCP transport, initializing
both the Transport client and server. The `x-pack/security` plugin provides
a secure version: `SecurityNetty4Transport` (with TLS and authentication).

A `Connection` between nodes is a pool of `Channels`, where each channel is a
non-blocking TCP connection (Java NIO terminology). Once a cluster is
discovered, a `Connection` (pool of `Channels`) is opened to every other node,
and every other node opens a `Connection` back. This results in two
`Connections` between any two nodes `(A→B and B→A)`. A node sends requests only
on the `Connection` it opens (acting as a client). The default pool is around 13

> `Channels`, divided into sub-pools for different purposes (e.g., ping,

Reviewer 2: Not sure if there is a convention here. I personally prefer `Connection`s over `Connections` since literal string seems to indicate corresponding java concept, e.g. class.
